### PR TITLE
DataWeave: Adding support for DataWeave language

### PR DESCRIPTION
**Pull request by machaval**
This pull request adds support for [dataweave](https://docs.mulesoft.com/mule-runtime/4.3/intro-dataweave2) language. 

```dataweave
%dw 2.0
var a = "123"
---
a
```

As seen in this snippet github already supports this language

### CODE REVIEW

File: components/prism-dataweave.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+(function (Prism) {
+    var keywords = /\b(?:match|input|output|ns|type|update|null|if|else|using|unless|at|is|as|case|do|fun|var|not|and|or)\b/;
+    
+    Prism.languages.dataweave = {	
+        "url": {
+            pattern:/(?:[A-z])+:\/\/(?:[A-z0-9/:\.\-_?=&])+|urn:(?:[A-z0-9:\-_\.?=&])+/
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:6]: (Please apply the fix for `A-z` here as well.)

```suggestion
            pattern:/\b[A-z]+:\/\/[A-z0-9/:\.\-_?=&]+|\burn:[A-z0-9:\-_\.?=&]+/
```

File: components/prism-dataweave.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        },
+        'property': {
+            pattern: /(?:[A-z0-9_]+#)?(?:"(?:\\.|[^\\"\r\n])*"|(?:[A-z0-9_]+))(?=\s*(?::|@))/,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:9]: (Please apply the fix for `A-z` here as well.)

```suggestion
            pattern: /(?:[A-z0-9_]+#)?(?:"(?:\\.|[^\\"\r\n])*"|[A-z0-9_]+)(?=\s*[:@])/,
```

File: tests/languages/dataweave/comment_feature.test [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	/* Block comment */
+	/**
+	* MultiLine Comment
+	**/
+
+----------------------------------------------------
+
+[
+	["comment", "// Line comment \"\"\t"],
+	["regex", "/* Block comment */"],
```
Comment (RunDevelopment) [inline tests/languages/dataweave/comment_feature.test:11]: Uhmm

File: components/prism-dataweave.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            pattern: /\/[^ ](?:[^\\\/\r\n]|\\[^\r\n])+\//,
+            greedy: true
+        },
+        'date': {
+            pattern: /\|[^ ](?:[^\\\|\r\n]|\\[^\r\n])+\|/,
+            greedy: true
+        },
+        'comment': [
+            {
+                pattern: /(^|[^\\])\/\*[\s\S]*?(?:\*\/|$)/,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:29]: This has to be greedy too.

Please add the snippet `/* "" */` to the tests as well.

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
Thank you for making this language definition @machaval!

There are a few things that have to be improved but overall solid work!

**Comment by machaval** [comment]
Thanks for all the feedback I will do the required changes ASAP

File: tests/languages/dataweave/comment_feature.test (hunk 3 above)
Comment (machaval) [reply tests/languages/dataweave/comment_feature.test:11]: Good catch!

**Comment by machaval** [comment]
All requested changes have been fixed now. Let me know if there is anything else you would like to be changed. Again thanks for all the comments you are the Regex master!

File: components/prism-dataweave.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        'property': {
+            pattern: /(?:[A-Za-z0-9_]+#)?(?:"(?:\\.|[^\\"\r\n])*"|[A-Za-z0-9_]+)(?=\s*[:@])/,
+            greedy: true
+        },
+        'string': {
+            pattern: /(["'`])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+            greedy: true
+        },
+        'mime-type': {
+            pattern: /(?:text|audio|video|application|multipart|image)\/(?:[A-Za-z-_+0-9])+/
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:15]:
```suggestion
            pattern: /\b(?:text|audio|video|application|multipart|image)\/[\w+-]+/
```

File: components/prism-dataweave.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+(function (Prism) {    
+    Prism.languages.dataweave = {	
+        'url': {
+            pattern:/\b(?:[A-Za-z])+:\/\/(?:[A-Za-z0-9/:\.\-_?=&])+|\burn:(?:[A-Za-z0-9:\-_\.?=&])+/
+        },
+        'property': {
+            pattern: /(?:[A-Za-z0-9_]+#)?(?:"(?:\\.|[^\\"\r\n])*"|[A-Za-z0-9_]+)(?=\s*[:@])/,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:7]:
```suggestion
            pattern: /(?:\w+#)?(?:"(?:\\.|[^\\"\r\n])*"|\w+)(?=\s*[:@])/,
```

File: components/prism-dataweave.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
+(function (Prism) {    
+    Prism.languages.dataweave = {	
+        'url': {
+            pattern:/\b(?:[A-Za-z])+:\/\/(?:[A-Za-z0-9/:\.\-_?=&])+|\burn:(?:[A-Za-z0-9:\-_\.?=&])+/
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:4]:
```suggestion
            pattern:/\b[A-Za-z]+:\/\/[\w/:.?=&-]+|\burn:[\w:.?=&-]+/
```

File: components/prism-dataweave.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        },
+        'string': {
+            pattern: /(["'`])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+            greedy: true
+        },
+        'mime-type': {
+            pattern: /(?:text|audio|video|application|multipart|image)\/(?:[A-Za-z-_+0-9])+/
+        },       
+        'date': {
+            pattern: /\|[^ ](?:[^\\|\r\n]|\\[^\r\n])+\|/,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:18]: Why `[^ ]`? Is it really just the space character that is disallowed?
Also, is `||a|` valid then?

(Same for `regex`)

File: components/prism-dataweave.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                greedy: true
+            },
+            {
+                pattern: /(^|[^\\:])\/\/.*/,
+                lookbehind: true,
+                greedy: true
+            }
+        ],
+        'regex': {
+            pattern: /\/[^ ](?:[^\\\/\r\n]|\\[^\r\n])+\//,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:34]: 1. Too bad `/a/` isn't a valid regex. Or is it? ;)
1. Is `/abc[/]/` not a valid regex? In languages like JS, character classes are parsed slightly differently. Is that not the case here?

```suggestion
            pattern: /\/[^ ](?:[^\\\/\r\n]|\\[^\r\n])*\//,
```

File: components/prism-dataweave.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            greedy: true
+        },
+        'function': /\b[A-Za-z_]\w*(?=\s*\()/i,
+        'number': /-?\b\d+(?:\.\d+)?(?:e[+-]?\d+)?\b/i,
+        'punctuation': /[{}[\];(),.:@]/,        
+        'operator': /<<|>>|->|[<>~=]=?|!=|--?-?|\+\+?|\!|\?/,
+        'boolean': /\b(?:true|false)\b/,
+        'keyword': {
+            pattern: /\b(?:match|input|output|ns|type|update|null|if|else|using|unless|at|is|as|case|do|fun|var|not|and|or)\b/
+        }
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:44]: Please simplify all rules that only have a `pattern` property.

```suggestion
        'keyword': /\b(?:match|input|output|ns|type|update|null|if|else|using|unless|at|is|as|case|do|fun|var|not|and|or)\b/
```

**Review by RunDevelopment — CHANGES_REQUESTED** [review]
Thank you for the quick response @machaval!

I left you a few comments again. A few regexes and rules can be simplified and I also have some questions.

File: components/prism-dataweave.js (hunk 9 above)
Comment (machaval) [reply components/prism-dataweave.js:34]: In dataweave if you want to put a `/` you need to  escape it `\/` `/abc[/]/` needs to be `/abc[\/]/`. Regarding spaces this is wrong. Also we dataweave doesn't support empty regex `//` is not a valid regex expression so shouldn't we keep the `+`

File: components/prism-dataweave.js (hunk 8 above)
Comment (machaval) [reply components/prism-dataweave.js]: I changed the regex to something clearer `/\|[\w\-:+-]+\|/` and added more tests where it shows all the possible values

File: components/prism-dataweave.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        },
+        'string': {
+            pattern: /(["'`])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+            greedy: true
+        },
+        'mime-type': {
+            pattern: /\b(?:text|audio|video|application|multipart|image)\/[\w+-]+/
+        },       
+        'date': {
+            pattern: /\|[\w\-:+-]+\|/,
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:18]:
```suggestion
            pattern: /\|[\w:+-]+\|/,
```

File: components/prism-dataweave.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            pattern: /(?:\w+#)?(?:"(?:\\.|[^\\"\r\n])*"|\w+)(?=\s*[:@])/,
+            greedy: true
+        },
+        'string': {
+            pattern: /(["'`])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+            greedy: true
+        },
+        'mime-type': {
+            pattern: /\b(?:text|audio|video|application|multipart|image)\/[\w+-]+/
+        },
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:16]: Just the regex is enough.

```suggestion
        'mime-type': /\b(?:text|audio|video|application|multipart|image)\/[\w+-]+/,
```

File: components/prism-dataweave.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
+(function (Prism) {    
+    Prism.languages.dataweave = {	
+        'url': {
+            pattern:/\b[A-Za-z]+:\/\/[\w/:.?=&-]+|\burn:[\w:.?=&-]+/
+        },
```
Comment (RunDevelopment) [inline components/prism-dataweave.js:5]: Same here.
```suggestion
        'url': /\b[A-Za-z]+:\/\/[\w/:.?=&-]+|\burn:[\w:.?=&-]+/,
```

**Review by RunDevelopment — APPROVED** [review]

**Comment by RunDevelopment** [comment]
Thank you for contributing @machaval!

**Pull request merged**